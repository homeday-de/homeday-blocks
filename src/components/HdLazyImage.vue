<template>
  <div
    v-if="background"
    ref="imageHolder"
    :class="{ isLoaded: imageLoaded }"
    class="lazy-image"
  />
  <img
    v-else
    ref="imageHolder"
    :class="{ isLoaded: imageLoaded }"
    class="lazy-image">
</template>

<script>
export default {
  name: 'HdLazyImage',
  props: {
    src: {
      type: String,
      default: '',
    },
    srcSmall: {
      type: String,
      default: '',
    },
    background: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      imageLoaded: false,
    };
  },
  watch: {
    src() {
      this.maybeLazyLoad();
    },
  },
  mounted() {
    this.maybeLazyLoad();
  },
  methods: {
    maybeLazyLoad() {
      this.setImageUrl('');

      if (!this.srcSmall) {
        this.setImageUrl(this.src);
        this.imageLoaded = true;
        return;
      }

      this.imageLoaded = false;
      this.setImageUrl(this.srcSmall);

      const image = new Image();

      image.onload = () => {
        this.imageLoaded = true;
        this.setImageUrl(this.src);
      };

      image.src = this.src;
    },
    setImageUrl(url) {
      if (!this.$refs.imageHolder) {
        return;
      }

      if (this.background) {
        this.$refs.imageHolder.style.backgroundImage = `url('${url}')`;
      } else {
        this.$refs.imageHolder.src = url;
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.lazy-image {
  width: 100%;
  filter: blur(4px);

  &.isLoaded {
    filter: blur(0);
  }
}
</style>
