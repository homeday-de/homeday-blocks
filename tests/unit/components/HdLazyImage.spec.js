import { mount } from '@vue/test-utils';
import HdLazyImage from '@/components/HdLazyImage.vue';
import ITEMS from '@/stories/mocks/GALLERY_ITEMS';

describe('HdLazyImage', () => {
  let wrapper;
  const sampleImage = ITEMS[2];

  beforeEach(() => {
    wrapper = mount(HdLazyImage, {
      propsData: {
        src: sampleImage.image,
        srcSmall: sampleImage.thumbnail,
        background: false,
      },
    });
  });

  it('renders component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
