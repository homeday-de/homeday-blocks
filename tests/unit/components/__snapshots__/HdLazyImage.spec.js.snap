// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdLazyImage renders component 1`] = `<img class="lazy-image" src="https://dummyimage.com/154x86?text=3+thumb+default+density">`;
